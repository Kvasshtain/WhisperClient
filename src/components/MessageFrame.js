import React from 'react'
import PropTypes from 'prop-types'
import { MessageTimestamp } from './MessageTimestamp'

function MessageFrame(props) {

    let message = props.message

    return (
        <div className="MessageFrame">
            <div>
                <p>{message.authorName}</p>
            </div>
            <div>
                <p>{message.authorEmail}</p>
            </div>
            <div>
                <p>{message.text}</p>
            </div>
            <MessageTimestamp time={message.time} />
        </div>
    )
}

MessageFrame.propTypes = {
    message: PropTypes.exact({
        chatId: PropTypes.number.isRequired,
        author: PropTypes.string.isRequired,
        time: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]).isRequired,
        text: PropTypes.string.isRequired,
        wasMessageReceived: PropTypes.bool.isRequired
    })
}

export { MessageFrame }