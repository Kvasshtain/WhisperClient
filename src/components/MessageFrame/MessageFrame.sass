$no-border: 0px solid white

.messageFrame
    display: inline-block
    margin: 10px
    border-radius: 10px
    background-color: #ffffff
    border: $no-border
    > div
        display: inline-block
        > div
            margin: 5px 10px 5px 10px
    .messageStatus
        display: inline-block
        width: 30px
        height: 30px