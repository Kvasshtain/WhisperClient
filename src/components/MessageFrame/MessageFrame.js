import React from 'react'
import PropTypes from 'prop-types'
import { MessageTimestamp } from './__MessageTimestamp/MessageFrame-MessageTimestamp'
import { AuthorNameLabel } from './__AuthorNameLabel/MessageFrame-AuthorNameLabel'
import { MessageTextField } from './__MessageTextField/MessageFrame-MessageTextField'

import './MessageFrame.sass'
import './_user/MessageFrame_user_another.sass'
import './_user/MessageFrame_user_current.sass'
import '../RotatingImage/_RotatingImage.sass'

import waitingIcon from '../../resources/BlueWaitingIcon.png'

class MessageFrame extends React.Component {
  renderAuthorName = () => {
    const { message, currentUserEmail } = this.props

    if (currentUserEmail === message.authorEmail) {
      return
    } else {
      return <AuthorNameLabel authorName={message.authorName} />
    }
  }

  renderMessageStatus() {
    const { hasServerReceivedMessage } = this.props

    if (!hasServerReceivedMessage) {
      return <div className="messageStatus">
        <img className="RotatingImage" src={waitingIcon}></img>
      </div>
    }
  }

  render() {
    const { message, currentUserEmail } = this.props
    let messageFrameClass = 'messageFrame'

    if (currentUserEmail === message.authorEmail) {
      messageFrameClass += ' messageFrame_user_current'
    } else {
      messageFrameClass += ' messageFrame_user_another'
    }

    return (
      <div className={messageFrameClass}>
        <div>
          {this.renderMessageStatus()}
        </div>
        <div>
          <div>{this.renderAuthorName()}</div>
          <div>
            <MessageTextField text={message.text} />
          </div>
          <div>
            <MessageTimestamp time={message.time} />
          </div>
        </div>
      </div>
    )
  }
}

MessageFrame.propTypes = {
  message: PropTypes.exact({
    _id: PropTypes.string,
    clientSideId: PropTypes.string,
    chatId: PropTypes.string.isRequired,
    authorName: PropTypes.string.isRequired,
    authorEmail: PropTypes.string.isRequired,
    time: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
      PropTypes.instanceOf(Date),
    ]).isRequired,
    text: PropTypes.string.isRequired,
    wasMessageReceived: PropTypes.bool,
  }),
  currentUserEmail: PropTypes.string,
  hasServerReceivedMessage: PropTypes.bool,
}

export { MessageFrame }
