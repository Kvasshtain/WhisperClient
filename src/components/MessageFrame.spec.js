import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json';

import { MessageFrame } from './MessageFrame'

describe('MessageFrame', () => {
    const mockMessage = {
        chatId: 0,
        author: 'captain Nemo',
        time: '2019-12-25T16:40:12.526Z',
        text: 'some message',
        wasMessageReceived: true,
    }

    it ('should render correctly', () => {
        const testComponent = shallow(<MessageFrame 
            message = { mockMessage }
        />)

        expect(shallowToJson(testComponent)).toMatchSnapshot();
      })

})