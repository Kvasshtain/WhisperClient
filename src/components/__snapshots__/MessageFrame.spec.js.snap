// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageFrame should render correctly 1`] = `
<div
  className="MessageFrame"
>
  <div>
    <p />
  </div>
  <div>
    <p />
  </div>
  <div>
    <p>
      some message
    </p>
  </div>
  <MessageTimestamp
    time="2019-12-25T16:40:12.526Z"
  />
</div>
`;
